In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [4]:
concrete_df = pd.read_csv("C:/Users/Lenovo/Documents/Data Set/concrete.csv")

In [5]:
concrete_df

cement   slag    ash  water  superplastic  coarseagg  fineagg  age  \
0      141.3  212.0    0.0  203.5           0.0      971.8    748.5   28   
1      168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14   
2      250.0    0.0   95.7  187.4           5.5      956.9    861.2   28   
3      266.0  114.0    0.0  228.0           0.0      932.0    670.0   28   
4      154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28   
...      ...    ...    ...    ...           ...        ...      ...  ...   
1025   135.0    0.0  166.0  180.0          10.0      961.0    805.0   28   
1026   531.3    0.0    0.0  141.8          28.2      852.1    893.7    3   
1027   276.4  116.0   90.3  179.6           8.9      870.1    768.3   28   
1028   342.0   38.0    0.0  228.0           0.0      932.0    670.0  270   
1029   540.0    0.0    0.0  173.0           0.0     1125.0    613.0    7   

      strength  
0        29.89  
1        23.51  
2        29.22  
3        45.85  
4        18.29  
...        ...  
1025     13.29  
1026     41.30  
1027     44.28  
1028     55.06  
1029     52.61  

[1030 rows x 9 columns]

In [6]:
concrete_df.head()

cement   slag    ash  water  superplastic  coarseagg  fineagg  age  \
0   141.3  212.0    0.0  203.5           0.0      971.8    748.5   28   
1   168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14   
2   250.0    0.0   95.7  187.4           5.5      956.9    861.2   28   
3   266.0  114.0    0.0  228.0           0.0      932.0    670.0   28   
4   154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28   

   strength  
0     29.89  
1     23.51  
2     29.22  
3     45.85  
4     18.29

In [7]:
concrete_df.tail()

cement   slag    ash  water  superplastic  coarseagg  fineagg  age  \
1025   135.0    0.0  166.0  180.0          10.0      961.0    805.0   28   
1026   531.3    0.0    0.0  141.8          28.2      852.1    893.7    3   
1027   276.4  116.0   90.3  179.6           8.9      870.1    768.3   28   
1028   342.0   38.0    0.0  228.0           0.0      932.0    670.0  270   
1029   540.0    0.0    0.0  173.0           0.0     1125.0    613.0    7   

      strength  
1025     13.29  
1026     41.30  
1027     44.28  
1028     55.06  
1029     52.61

In [8]:
concrete_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cement        1030 non-null   float64
 1   slag          1030 non-null   float64
 2   ash           1030 non-null   float64
 3   water         1030 non-null   float64
 4   superplastic  1030 non-null   float64
 5   coarseagg     1030 non-null   float64
 6   fineagg       1030 non-null   float64
 7   age           1030 non-null   int64  
 8   strength      1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


In [9]:
concrete_df.isnull().sum()

cement          0
slag            0
ash             0
water           0
superplastic    0
coarseagg       0
fineagg         0
age             0
strength        0
dtype: int64

In [10]:
concrete_df.describe()

cement         slag          ash        water  superplastic  \
count  1030.000000  1030.000000  1030.000000  1030.000000   1030.000000   
mean    281.167864    73.895825    54.188350   181.567282      6.204660   
std     104.506364    86.279342    63.997004    21.354219      5.973841   
min     102.000000     0.000000     0.000000   121.800000      0.000000   
25%     192.375000     0.000000     0.000000   164.900000      0.000000   
50%     272.900000    22.000000     0.000000   185.000000      6.400000   
75%     350.000000   142.950000   118.300000   192.000000     10.200000   
max     540.000000   359.400000   200.100000   247.000000     32.200000   

         coarseagg      fineagg          age     strength  
count  1030.000000  1030.000000  1030.000000  1030.000000  
mean    972.918932   773.580485    45.662136    35.817961  
std      77.753954    80.175980    63.169912    16.705742  
min     801.000000   594.000000     1.000000     2.330000  
25%     932.000000   730.950000     7.000000    23.710000  
50%     968.000000   779.500000    28.000000    34.445000  
75%    1029.400000   824.000000    56.000000    46.135000  
max    1145.000000   992.600000   365.000000    82.600000

In [11]:
concrete_df.shape

(1030, 9)

In [15]:
X = concrete_df.drop(["strength"],axis = 1)

In [16]:
X

cement   slag    ash  water  superplastic  coarseagg  fineagg  age
0      141.3  212.0    0.0  203.5           0.0      971.8    748.5   28
1      168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14
2      250.0    0.0   95.7  187.4           5.5      956.9    861.2   28
3      266.0  114.0    0.0  228.0           0.0      932.0    670.0   28
4      154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28
...      ...    ...    ...    ...           ...        ...      ...  ...
1025   135.0    0.0  166.0  180.0          10.0      961.0    805.0   28
1026   531.3    0.0    0.0  141.8          28.2      852.1    893.7    3
1027   276.4  116.0   90.3  179.6           8.9      870.1    768.3   28
1028   342.0   38.0    0.0  228.0           0.0      932.0    670.0  270
1029   540.0    0.0    0.0  173.0           0.0     1125.0    613.0    7

[1030 rows x 8 columns]

In [19]:
Y = concrete_df[["strength"]]

In [20]:
Y

strength
0        29.89
1        23.51
2        29.22
3        45.85
4        18.29
...        ...
1025     13.29
1026     41.30
1027     44.28
1028     55.06
1029     52.61

[1030 rows x 1 columns]

In [21]:
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)

In [22]:
X

array([[-1.33901711,  1.60144087, -0.84714393, ..., -0.01439767,
        -0.31296991, -0.27973311],
       [-1.07479007, -0.36754132,  1.09607803, ...,  1.38814131,
         0.28226038, -0.50146528],
       [-0.29838379, -0.85688789,  0.64896501, ..., -0.20612089,
         1.09337085, -0.27973311],
       ...,
       [-0.04564488,  0.4882354 ,  0.56454507, ..., -1.32300514,
        -0.06589318, -0.27973311],
       [ 0.58237302, -0.41624406, -0.84714393, ..., -0.52651741,
        -1.29254178,  3.55306569],
       [ 2.47791487, -0.85688789, -0.84714393, ...,  1.9568773 ,
        -2.00382326, -0.61233136]])

In [23]:
from sklearn import preprocessing
Y = preprocessing.StandardScaler().fit(Y).transform(Y)

In [24]:
Y

array([[-0.35501811],
       [-0.73710825],
       [-0.39514356],
       ...,
       [ 0.50678082],
       [ 1.15238141],
       [ 1.005654  ]])

In [25]:
from sklearn.model_selection import train_test_split

In [26]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=200)
print ('Train set:', X_train.shape,  Y_train.shape)
print ('Test set:', X_test.shape,  Y_test.shape)

Train set: (824, 8) (824, 1)
Test set: (206, 8) (206, 1)


In [27]:
X_train

array([[-1.07479007, -0.36754132,  1.09607803, ...,  1.38814131,
         0.28226038, -0.50146528],
       [-1.14084683,  1.34633129,  1.46658777, ..., -1.73604643,
        -0.40655958, -0.27973311],
       [-0.41135043,  0.9845395 , -0.84714393, ...,  0.87087464,
        -0.49016635, -0.27973311],
       ...,
       [-0.30699989, -0.85688789,  0.69742831, ...,  0.19147961,
         1.44027655, -0.50146528],
       [ 0.3143166 ,  0.82799498,  0.92254815, ..., -1.33587247,
        -1.0404736 , -0.27973311],
       [ 0.14007993, -0.85688789, -0.84714393, ...,  1.52453501,
        -0.05341456, -0.61233136]])

In [28]:
X_test

array([[-1.18871405,  1.87974224, -0.84714393, ..., -0.48276848,
         0.09508105, -0.27973311],
       [-1.25572815,  1.88901896, -0.84714393, ...,  1.24016701,
        -1.23264439, -0.27973311],
       [-0.4927247 , -0.85688789,  1.00071476, ...,  0.71003302,
        -0.19941445,  0.86060372],
       ...,
       [-0.3979476 ,  3.31067512, -0.84714393, ..., -0.40299104,
        -1.36366993, -0.61233136],
       [ 0.41962448, -0.85688789, -0.84714393, ...,  1.15910284,
         0.11754257, -0.27973311],
       [ 0.9940311 , -0.85688789, -0.84714393, ..., -0.08902818,
        -0.13202988, -0.61233136]])

In [29]:
Y_train

array([[-7.37108252e-01],
       [-1.23248695e-01],
       [-5.33486546e-01],
       [ 2.37650720e+00],
       [-4.78388907e-01],
       [-1.58752832e+00],
       [ 2.80171723e+00],
       [ 2.59440335e-01],
       [-6.20924102e-01],
       [-1.75950784e-01],
       [ 3.13340199e-01],
       [ 7.18188063e-01],
       [-8.83835658e-01],
       [-1.02217864e+00],
       [-1.06949074e+00],
       [ 5.06780820e-01],
       [-7.36509365e-01],
       [ 1.90937505e+00],
       [-2.28053986e-01],
       [-1.66368586e-01],
       [ 1.63628240e+00],
       [ 3.02560226e-01],
       [-1.44808641e-01],
       [ 3.50471216e-01],
       [-1.06529853e+00],
       [-1.00061870e+00],
       [-1.52594177e-01],
       [-1.94626186e+00],
       [ 4.64259817e-01],
       [ 3.78618923e-01],
       [-1.75761233e+00],
       [ 2.86390267e-01],
       [ 9.71418576e-02],
       [-4.71801146e-01],
       [-1.38270884e+00],
       [-9.03598942e-01],
       [-1.35935223e+00],
       [ 7.72686814e-01],
       [-2.8

In [30]:
Y_test

array([[-1.29237569e-01],
       [ 9.65429702e-02],
       [ 3.01961339e-01],
       [ 5.02588609e-01],
       [-3.61006982e-01],
       [-5.49755350e-02],
       [ 4.60067605e-01],
       [-2.67580552e-01],
       [ 3.95387769e-01],
       [-1.17010382e+00],
       [-1.42882317e+00],
       [-8.40715768e-01],
       [ 8.81085428e-01],
       [ 2.59440335e-01],
       [ 2.14893000e+00],
       [ 1.25059894e+00],
       [ 2.32490404e-01],
       [ 2.12497450e+00],
       [-1.24496474e+00],
       [ 1.59795361e+00],
       [ 7.38550233e-01],
       [-3.88055760e-02],
       [-9.86245399e-01],
       [ 8.69607722e-02],
       [ 1.09488822e+00],
       [-1.27431023e+00],
       [-8.95214518e-01],
       [ 2.60408440e+00],
       [ 6.86447032e-01],
       [ 1.68898449e+00],
       [ 1.02362062e+00],
       [ 1.46559950e+00],
       [-7.95299173e-02],
       [ 1.84349744e+00],
       [-6.81510572e-02],
       [ 3.13340199e-01],
       [-8.42512430e-01],
       [ 1.01823064e+00],
       [-9.2

# Multi Linear Regression Model

In [46]:
from sklearn import linear_model
regr = linear_model.LinearRegression()
regr.fit(X_train, Y_train)#training func question + answers
# The coefficients
print ('Intercept: ',regr.intercept_)
print ('Coefficient : ',regr.coef_)

Intercept:  [-0.00664992]
Coefficient :  [[ 0.76306074  0.53781737  0.34583184 -0.22882221  0.06041347  0.05372764
   0.09716039  0.42315693]]


In [47]:
yhat = clf.predict(X_test) #question 
yhat [0:5]

array([323.90666181, 359.77470046, 345.03595327, 354.29778162,
       371.45936353])

In [50]:
regr.intercept_

array([-0.00664992])

In [51]:
regr.coef_

array([[ 0.76306074,  0.53781737,  0.34583184, -0.22882221,  0.06041347,
         0.05372764,  0.09716039,  0.42315693]])

In [52]:
regr.coef_[0][0]

0.7630607394912605

In [53]:
regr.coef_[0][1]

0.5378173673339044

In [54]:
regr.coef_[0][2]

0.34583183662349165

In [55]:
y_pred = regr.predict(X_test)

In [56]:
y_pred

array([[-0.50541801],
       [-0.2724768 ],
       [-0.26166731],
       [ 0.32859334],
       [ 1.04732279],
       [ 0.78507292],
       [-0.26366244],
       [-0.55655374],
       [ 0.52962561],
       [-0.76295636],
       [-0.79366127],
       [-0.11389582],
       [ 0.51345118],
       [-0.07426859],
       [ 1.42717894],
       [ 1.33688772],
       [-0.41355436],
       [ 1.08693431],
       [-0.96981741],
       [ 1.75911148],
       [ 0.11672317],
       [-0.26348065],
       [-0.59660297],
       [-0.65937596],
       [ 0.94825273],
       [-0.29042524],
       [-1.02980628],
       [ 1.50915807],
       [ 0.63325212],
       [ 0.73926479],
       [-0.45578163],
       [ 1.29398647],
       [ 0.18640063],
       [ 1.35759609],
       [-0.33668513],
       [ 0.55127699],
       [-0.96582099],
       [ 0.52446842],
       [-0.57493751],
       [-0.26493585],
       [ 0.17581027],
       [ 1.59528474],
       [-0.71702797],
       [ 2.29224226],
       [ 1.13162513],
       [ 0

In [57]:
from sklearn.metrics import r2_score 

In [58]:
print(f"R2 Score : {r2_score(Y_test,y_pred)*100} % ")

R2 Score : 58.06420271412136 % 


In [59]:
print(f"Mean absolute error: {np.mean(np.absolute(y_pred - Y_test))} ")#pred - actual

Mean absolute error: 0.502018043469937 


In [60]:
print("Residual sum of squares (MSE): %.2f" % np.mean((y_pred - Y_test) **2))

Residual sum of squares (MSE): 0.40


In [61]:
accuracy = print(f"accuracy : {r2_score(Y_test,y_pred)*100} % ")

accuracy : 58.06420271412136 % 


# Decision Tree

In [63]:
from sklearn.tree import DecisionTreeRegressor
concrete_df = DecisionTreeRegressor(criterion="squared_error", max_depth = 4)
concrete_df # it shows the default parameters

DecisionTreeRegressor(max_depth=4)

In [64]:
concrete_df.fit(X_train,Y_train)

DecisionTreeRegressor(max_depth=4)

# Prediction

In [66]:
y_pred = concrete_df.predict(X_test)

In [67]:
y_pred

array([-0.37574447, -0.37574447, -0.08992693, -0.37574447, -0.08992693,
        0.66383903,  0.54325664, -0.08992693, -0.12608343, -0.82698128,
       -1.38935394, -1.38935394, -0.08992693, -0.08992693,  1.35202868,
        1.71160913, -0.08992693,  1.71160913, -1.38935394,  1.71160913,
        0.54325664,  0.32531795, -1.38935394,  0.54325664,  0.66383903,
       -0.82698128, -0.08992693,  1.71160913,  1.06005693,  0.32531795,
        0.32531795,  1.71160913, -0.08992693,  1.71160913, -0.08992693,
        0.32531795, -0.08992693,  0.54325664, -1.38935394,  0.54325664,
        0.54325664,  1.71160913, -0.72862401,  0.32531795,  1.06005693,
        0.54325664, -0.8770279 ,  1.71160913, -1.26514264,  0.54325664,
        0.54325664, -0.37574447, -0.37574447,  1.71160913,  0.32531795,
       -0.72862401,  0.54325664,  1.06005693,  0.54325664, -0.08992693,
       -0.08992693,  0.32531795, -1.26514264,  0.54325664, -0.08992693,
       -1.38935394, -0.08992693, -0.26979383,  0.54325664, -0.08

In [68]:
print (y_pred [0:5])#predicted by the ml model
print (Y_test [0:5])#actual values we have

[-0.37574447 -0.37574447 -0.08992693 -0.37574447 -0.08992693]
[[-0.12923757]
 [ 0.09654297]
 [ 0.30196134]
 [ 0.50258861]
 [-0.36100698]]


In [69]:
Y_test

array([[-1.29237569e-01],
       [ 9.65429702e-02],
       [ 3.01961339e-01],
       [ 5.02588609e-01],
       [-3.61006982e-01],
       [-5.49755350e-02],
       [ 4.60067605e-01],
       [-2.67580552e-01],
       [ 3.95387769e-01],
       [-1.17010382e+00],
       [-1.42882317e+00],
       [-8.40715768e-01],
       [ 8.81085428e-01],
       [ 2.59440335e-01],
       [ 2.14893000e+00],
       [ 1.25059894e+00],
       [ 2.32490404e-01],
       [ 2.12497450e+00],
       [-1.24496474e+00],
       [ 1.59795361e+00],
       [ 7.38550233e-01],
       [-3.88055760e-02],
       [-9.86245399e-01],
       [ 8.69607722e-02],
       [ 1.09488822e+00],
       [-1.27431023e+00],
       [-8.95214518e-01],
       [ 2.60408440e+00],
       [ 6.86447032e-01],
       [ 1.68898449e+00],
       [ 1.02362062e+00],
       [ 1.46559950e+00],
       [-7.95299173e-02],
       [ 1.84349744e+00],
       [-6.81510572e-02],
       [ 3.13340199e-01],
       [-8.42512430e-01],
       [ 1.01823064e+00],
       [-9.2

In [71]:
from sklearn.metrics import r2_score 

In [72]:
print(f"R2 Score : {r2_score(Y_test,y_pred)*100} % ")

R2 Score : 69.62798048370925 % 


In [74]:
accuracy = (f"accuracy : {r2_score(Y_test,y_pred)*100} % ")

In [75]:
accuracy

'accuracy : 69.62798048370925 % '

# Random Forest

In [76]:
from sklearn.ensemble import RandomForestRegressor
Regressor = RandomForestRegressor(n_estimators = 400, criterion = 'squared_error') #400 decision trees
Regressor.fit(X_train,Y_train)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_2164\1037969464.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  Regressor.fit(X_train,Y_train)


RandomForestRegressor(n_estimators=400)

In [77]:
# Predicting the test set results
Y_Pred = Regressor.predict(X_test)#examination only question are given

In [78]:
Y_Pred

array([-0.4096037 , -0.07478224,  0.1951378 ,  0.25307566, -0.4494896 ,
        0.62248841,  0.57314054, -0.06994023, -0.00516458, -1.05651485,
       -1.45831238, -0.90366482,  0.6997034 ,  0.28844745,  1.91617541,
        1.39468076,  0.54526532,  2.11683412, -1.35144991,  1.48511125,
        0.30847424,  0.02232436, -1.19415364, -0.47620147,  0.69905025,
       -1.20197212, -0.89432367,  2.59182068,  0.83058682,  0.57249973,
        0.90523406,  1.47487178, -0.04381976,  1.61539022, -0.0284029 ,
        0.26685306, -0.56549258,  0.80868593, -1.29695266,  0.24201122,
        0.64740107,  2.13190662, -0.72332785,  1.44145536,  1.66140423,
        1.29508279, -0.619635  ,  2.07121538, -1.26638545,  0.90917175,
       -0.66585712,  0.04161452, -0.59326449,  2.59182068, -0.1328803 ,
       -0.86708927,  0.8917037 ,  1.5076676 ,  0.83118314,  0.7398588 ,
       -0.5850253 ,  0.24112786, -0.80592939,  0.99587267, -0.16805445,
       -1.13284854,  0.99314774, -0.09842856,  0.03698362,  0.70

In [81]:
from sklearn.metrics import r2_score 

In [82]:
print(f"R2 Score : {r2_score(Y_test,Y_Pred)*100} % ")

R2 Score : 91.44258063532135 % 


In [83]:
print(f"Mean absolute error: {np.mean(np.absolute(Y_Pred - Y_test))} ")#pred - actual

Mean absolute error: 1.0707031747156566 


In [84]:
print("Residual sum of squares (MSE): %.2f" % np.mean((Y_Pred - Y_test) **2))

Residual sum of squares (MSE): 1.80


# Ans : Random Forest gives the best accuracy that is 91.44%